src/libsyntax/tokenstream.rs

```diff
@@ -230,21 +229,25 @@ impl PartialEq<TokenStream> for TokenStream {
 
 impl TokenStream {
     pub fn len(&self) -> usize {
-        if let TokenStream::Stream(ref slice) = self {
+        if let Some(ref slice) = self.0 {
             slice.len()
         } else {
             0
         }
     }
 
     pub fn empty() -> TokenStream {
-        TokenStream::Empty
+        TokenStream(None)
     }
 
     pub fn is_empty(&self) -> bool {
-        match self {
-            TokenStream::Empty => true,
-            _ => false,
+        match self.0 {
+            None => true,
+            Some(ref stream) => {
+                // An empty stream should be represented as `None`, not as an empty `Vec`.
+                debug_assert!(stream.len() > 0);
```

Comment:
Hmm, do we have operations that remove trees from the stream's vector?
It's correct that we should not create *new* empty streams as `Some`, but perhaps previously non-empty streams can turn into empty?

Reply:
Not that I know of. But if it does happen, the old code for this function would silently do the wrong thing, while the new code will assert. I could change it to check for a length of zero in the `Some` case.

Reply:
Yes, could you use the correct `Some(ref stream) => stream.is_empty()`, that would be more future-proof.